expression/simple_rewriter.go

```diff
@@ -0,0 +1,455 @@
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/parser"
+	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+)
+
+type simpleRewriter struct {
+	tbl *model.TableInfo
+
+	stack *ExprStack
+	err   error
+	ctx   sessionctx.Context
+}
+
+// ParseSimpleExpr parses simple expression string to Expression.
+// The expression string must only reference the column in table Info.
+func ParseSimpleExpr(ctx sessionctx.Context, exprStr string, tableInfo *model.TableInfo) (Expression, error) {
+	exprStr = fmt.Sprintf("select %s", exprStr)
+	stmts, err := parser.New().Parse(exprStr, "", "")
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	expr := stmts[0].(*ast.SelectStmt).Fields.Fields[0].Expr
+	rewriter := &simpleRewriter{tbl: tableInfo, ctx: ctx, stack: new(ExprStack)}
+	expr.Accept(rewriter)
+	if rewriter.err != nil {
+		return nil, errors.Trace(rewriter.err)
+	}
+	return rewriter.stack.Pop(), nil
+}
+
+func (sr *simpleRewriter) rewriteColumn(cn *ast.ColumnNameExpr) (*Column, error) {
+	cols := sr.tbl.Columns
+	for i, col := range cols {
+		if col.Name.L == cn.Name.Name.L {
+			eCol := &Column{
+				FromID:      1,
+				ColName:     col.Name,
+				OrigTblName: sr.tbl.Name,
+				DBName:      model.NewCIStr(sr.ctx.GetSessionVars().CurrentDB),
+				TblName:     sr.tbl.Name,
+				RetType:     &col.FieldType,
+				ID:          col.ID,
+				Position:    col.Offset,
+				Index:       i,
+			}
+			return eCol, nil
+		}
+	}
+	return nil, errBadField.Gen(cn.Name.Name.O, "expression")
+}
+
+func (sr *simpleRewriter) Enter(inNode ast.Node) (ast.Node, bool) {
+	return inNode, false
+}
+
+func (sr *simpleRewriter) Leave(originInNode ast.Node) (retNode ast.Node, ok bool) {
+	switch v := originInNode.(type) {
+	case *ast.ColumnNameExpr:
+		column, err := sr.rewriteColumn(v)
+		if err != nil {
+			sr.err = err
+			return originInNode, false
+		}
+		sr.stack.Push(column)
+	case *ast.ValueExpr:
+		value := &Constant{Value: v.Datum, RetType: &v.Type}
+		sr.stack.Push(value)
+	case *ast.FuncCallExpr:
+		sr.funcCallToExpression(v)
+	case *ast.FuncCastExpr:
+		arg := sr.stack.Pop()
+		sr.err = CheckArgsNotMultiColumnRow(arg)
+		if sr.err != nil {
+			return retNode, false
+		}
+		sr.stack.Push(BuildCastFunction(sr.ctx, arg, v.Tp))
+	case *ast.BinaryOperationExpr:
+		sr.binaryOpToExpression(v)
+	case *ast.UnaryOperationExpr:
+		sr.unaryOpToExpression(v)
+	case *ast.BetweenExpr:
+		sr.betweenToExpression(v)
+	case *ast.IsNullExpr:
+		sr.isNullToExpression(v)
+	case *ast.IsTruthExpr:
+		sr.isTrueToScalarFunc(v)
+	case *ast.PatternLikeExpr:
+		sr.likeToScalarFunc(v)
+	case *ast.PatternRegexpExpr:
+		sr.regexpToScalarFunc(v)
+	case *ast.PatternInExpr:
+		if v.Sel == nil {
+			sr.inToExpression(len(v.List), v.Not, &v.Type)
+		}
+	case *ast.RowExpr:
+		sr.rowToScalarFunc(v)
+	case *ast.ParenthesesExpr:
+	case *ast.ColumnName:
```

Comment:
why ignore this?

Reply:
It will be visited because it's a `ast.Node`, but we handle it in `*ast.ColumnNameExpr`.
If not list it here, we will get an unknown type error.